tests/unit/test_oauth_provider.py

```diff
@@ -0,0 +1,200 @@
+            "token_endpoint_auth_method": "client_secret_basic",
+        },
+    )
+    relation_data = harness.get_relation_data(relation_id, harness.model.app.name)
+
+    client_secret_id = relation_data.pop("client_secret_id")
+    secret = harness.model.get_secret(id=client_secret_id)
+
+    assert len(harness.charm.events) == 1
+    assert isinstance(harness.charm.events[0], ClientCreatedEvent)
+    assert secret.get_content()[CLIENT_SECRET_FIELD] == CLIENT_SECRET
+    assert relation_data == {
+        "authorization_endpoint": "https://example.oidc.com/oauth2/auth",
+        "introspection_endpoint": "https://example.oidc.com/admin/oauth2/introspect",
+        "issuer_url": "https://example.oidc.com",
+        "jwks_endpoint": "https://example.oidc.com/.well-known/jwks.json",
+        "scope": "openid profile email phone",
+        "token_endpoint": "https://example.oidc.com/oauth2/token",
+        "userinfo_endpoint": "https://example.oidc.com/userinfo",
+        "client_id": CLIENT_ID,
+    }
+
+
+def test_client_changed(harness):
+    relation_id = harness.add_relation("oauth", "requirer")
+    harness.add_relation_unit(relation_id, "requirer/0")
+    harness.update_relation_data(
+        relation_id,
+        "requirer",
+        {
+            "redirect_uri": "https://oidc-client.com/callback",
+            "scope": "openid email",
+            "grant_types": '["authorization_code"]',
+            "audience": "[]",
+            "token_endpoint_auth_method": "client_secret_basic",
+        },
+    )
+    assert len(harness.charm.events) == 1
+    assert isinstance(harness.charm.events[0], ClientCreatedEvent)
+
+    redirect_uri = "https://oidc-client.com/callback2"
+    harness.update_relation_data(
+        relation_id,
+        "requirer",
+        {
+            "redirect_uri": redirect_uri,
+            "scope": "openid email",
+            "grant_types": '["authorization_code"]',
+            "audience": "[]",
+            "token_endpoint_auth_method": "client_secret_basic",
+        },
+    )
+
+    assert len(harness.charm.events) == 2
+    assert isinstance(harness.charm.events[1], ClientChangedEvent)
+    assert harness.charm.events[1].redirect_uri == redirect_uri
+
+
+def test_on_client_config_deleted_event_emitted(harness):
+    relation_id = harness.add_relation("oauth", "requirer")
+    harness.add_relation_unit(relation_id, "requirer/0")
+    harness.update_relation_data(
+        relation_id,
+        "requirer",
+        {
+            "redirect_uri": "https://oidc-client.com/callback",
+            "scope": "openid email",
+            "grant_types": '["authorization_code"]',
+            "audience": "[]",
+            "token_endpoint_auth_method": "client_secret_basic",
+        },
+    )
+    assert len(harness.charm.events) == 1
+    assert isinstance(harness.charm.events[0], ClientCreatedEvent)
+
+    harness.remove_relation(relation_id)
+
+    assert len(harness.charm.events) == 2
```

Comment:
Do we really want to num of events to be 2? We don't really care right? What we care about in this test is that the `ClientDeletedEvent` was triggered.

Reply:
Well it's a bit more nuanced than this. Take this example

```python
def test_xyz():
    """
    Given: Bank Account with initial balance of 10$
    When: Transfer 5$
    Then: Balance is 15$
    """
    bank_account = BankAccount(balance=10)
    
    bank_account.transfer(5)

    assert bank_account.balance == 15
```

It's not so much that this test is bad. Indeed if the transfer is of 5$, we expect the balance to be 15$. The problem is that it's not asserting what we really want to assert which is that the account balance is the sum of the initial balance and the transfer amount. So a better way to write it is:

```python
def test_xyz():
    """
    Given: Bank Account with initial balance
    When: Transfer amount
    Then: Balance is sum of initial balance and transfer amount
    """
    initial_balance = 10
    transfer_amount = 5
    bank_account = BankAccount(balance=initial_balance)
    
    bank_account.transfer(transfer_amount)

    assert bank_account.balance == initial_balance + transfer_amount
```